rfcs/ocamlib.md

```diff
@@ -0,0 +1,548 @@
+be a mix of archive using `META` files and others using the new
+mecanism it must take care to resolve the dependencies of the latter
+according to the `OCAMLPATH` semantics and specify `-noautoliblink` so
+that no double linking occurs.
+
+## `dune` support
+
+Currently, Dune reads `META` files of installed libraries that were
+not built by Dune and generates `META` files for all the libraries it
+installs.
+
+Once the compiler supports this proposal, Dune will start installing
+library artifacts using the convention described in this document. It
+will also pass the `-lib-require` field when assembling library
+archives. This will be a non-breaking change since from the point of
+view of users the naming of artifacts is a implementation detail of
+Dune. Dune will only enforce the new convention for the libraries it
+installs but will not assume it. This way, users of Dune have nothing
+to do and projects that do not use Dune do not have to be upgraded all
+at once.
+
+Starting from Dune 3, Dune will assume that all OCaml libraries follow
+the convention described in this document and will stop reading `META`
+files. This means that it will no longer be possible to consume legacy
+libraries in Dune. On the plus side, it will simplify a bit the Dune
+code base.
+
+Starting from Dune 4, Dune will no longer generate `META` files.
+
+In newer `dune` configuration files, Dune will require library names
+to use `/` rather than `.`. However, to ensure backward compatibility
+Dune will transparently translate `.` to `/` in library names when
+reading older `dune` configuration files. This will ensure that newer
+projects use the new naming conventions without breaking the build of
+projects that haven't been updated in a while.
+
+
+## Unresolved issues 
+
+### Compilation phase includes
+
+At the moment the proposal indicates that during the compilation phase
+use of a library via `-lib LIB` simply adds `LIB`'s directory library
+to the includes for the compilation. The include directories of the
+libraries `LIB` depends on are not added. The reason is that doing the
+latter leads to dependency underspecification (see
+[here](https://github.com/ocaml/opam-repository/pull/13803) for a real
+example).
+
+The rationale is that if the library `L` does not make its usage of
+another library `LD` abstract and that it leaks in `L`'s interface
+then concretly the client of `L` also need to depend on `LD`.
+
+The problem of this approach is that this can lead to errors that are
+difficult to diagnose and that this notion seems to be currently ill
+defined.
+
+Here are two approaches that have been proposed to alleviate the problem: 
+
+1. Make `cmi` files self-contained by expanding type aliases, module 
+   aliases and module type aliases.
+2. Add a `--lib-hidden LIB` option that tells the compiler to also 
+   look into `LIB` for `cmis` but without directly exposing the 
+   declarations to the source files that are being compiled.
```

Comment:
The approach I had in mind in [my comment](https://gist.github.com/dbuenzli/a78131f54580212986713ef3e9b313e8#gistcomment-3083085), and I believe also in Leo's [followup](https://gist.github.com/dbuenzli/a78131f54580212986713ef3e9b313e8#gistcomment-3083103) was rather to add a `--hiden-cmi foo.cmi` or `-I-hidden foo/` option to the compiler (the later, more global approach allows us to decide to also exfiltrate certain information from `.cmx` files for example, or we may also need a `--hiden-cmx foo.cmx` option), and then have transitive dependencies of `--lib` arguments be turned into `-I-hidden <transdeps>` on the command-line instead of nothing at all. The intent to hide would be implicit in the use of the `-lib` flag (for its transitive dependencies that are not also direct dependencies), instead of having to be passed explicitly by the user. (It's not clear to me whether you intend the `--lib-hidden` flag to be passed by the user explicitly, or by the compiler implicitly from `-lib` arguments).

In the presence of this logic, indeed the compiler needs to know of transitive dependencies also during the compilation file, and indeed currently that means reading `.cma` file, which was a worry of Alain. I believe that the existence of this design choice (not yet resolved) might justify the idea of storing library dependencies in other intermediate files than `.cma` in addition, here in both `.cmi` and `.cmx`. Having the information in those files (is not very hard to implement and) extends our design/decision space for what we want to do on transitive dependencies, so it sounds like a safer design to me.